### PR TITLE
Adds database pagination module to helpers crate

### PR DESCRIPTION
**Pull request by code-mm**
This adds a database pagination module to the helpers crate. The pagination feature can split the results in pages and return the number of pages.

```
<table>
    .filter(<some-filter>)
    .paginate(<page>)
    .per_page(<results-per-page>)
    .load_and_count_pages(<db-connection>)
```
Returns a `<Vec<T>, i64>`, where T is the model type.

### CODE REVIEW

File: helpers/src/db/pagination.rs [hunk 1]
```diff
@@ -0,0 +1,69 @@
+use diesel::pg::Pg;
+use diesel::prelude::*;
+use diesel::query_builder::*;
+use diesel::query_dsl::methods::LoadQuery;
+use diesel::sql_types::BigInt;
+use diesel::QueryId;
+
+const DEFAULT_PER_PAGE: i64 = 10;
+
+pub trait Paginate: Sized {
```
Comment (Urhengulas) [inline helpers/src/db/pagination.rs:10]: Is `Sized` really the only prerequisite? I would assume we should restrict it to `LoadQuery<PgConnection, (U, i64)>` (from line 38), because else the most important method won't work.

File: helpers/src/db/pagination.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    type SqlType = (T::SqlType, BigInt);
+}
+
+impl<T> RunQueryDsl<PgConnection> for Paginated<T> {}
+
+impl<T> QueryFragment<Pg> for Paginated<T>
+where
+    T: QueryFragment<Pg>,
+{
+    fn walk_ast(&self, mut out: AstPass<Pg>) -> QueryResult<()> {
```
Comment (Urhengulas) [inline helpers/src/db/pagination.rs:59]: Can this be simplified? And please add a brief commend explaining what it is doing.

File: helpers/src/db/pagination.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        out.push_sql("SELECT *, COUNT(*) OVER () FROM (");
+        self.query.walk_ast(out.reborrow())?;
+        out.push_sql(") t LIMIT ");
+        out.push_bind_param::<BigInt, _>(&self.per_page)?;
+        out.push_sql(" OFFSET ");
```
Comment (Urhengulas) [inline helpers/src/db/pagination.rs:64]: Since you are using plain SQL `LIMIT` and `OFFSET` I wanted to mention that I planned to use the the methods for that:
* [`.limit(limit: i64)`](https://docs.rs/diesel/1.4.5/diesel/query_dsl/trait.QueryDsl.html#method.limit)
* [`.offset(offset: i64)`](https://docs.rs/diesel/1.4.5/diesel/query_dsl/trait.QueryDsl.html#method.offset)

Maybe you can also use that for the trait to make it easier understandable?

**Review by Urhengulas — CHANGES_REQUESTED** [review]
Seems like a very useful feature!

Can you please provide a usage example to the doc-comment, so I can try it out with my code?

File: helpers/src/db/pagination.rs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    page: i64,
+    per_page: i64,
+}
+
+impl<T> Paginated<T> {
+    pub fn per_page(self, per_page: i64) -> Self {
+        Paginated { per_page, ..self }
+    }
+
+    pub fn load_and_count_pages<U>(self, conn: &PgConnection) -> QueryResult<(Vec<U>, i64)>
```
Comment (Urhengulas) [inline helpers/src/db/pagination.rs:36]: And shouldn't this return a `QueryResult<(U, i64)>`, because I only want one page and not all pages?

**Review by Urhengulas — CHANGES_REQUESTED** [review]

File: helpers/src/db/pagination.rs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+#[derive(Debug, Clone, Copy, QueryId)]
+pub struct Paginated<T> {
+    query: T,
+    page: i64,
+    per_page: i64,
+}
+
+impl<T> Paginated<T> {
+    pub fn per_page(self, per_page: i64) -> Self {
+        Paginated { per_page, ..self }
```
Comment (Urhengulas) [inline helpers/src/db/pagination.rs:33]: This somehow reads a bit weirdly..

What do you think about
* 1
    ```rust
    pub fn per_page(self, per_page: i64) -> Self {
            self.per_page = per_page;
            self
    }
    ```
* 2
    ```rust
    pub fn per_page(&mut self, per_page: i64) -> Self {
            self.per_page = per_page;
            self.clone()
    }
    ```
    * this would also enable to use it in-place and not only trailing

**Review by Urhengulas — CHANGES_REQUESTED** [review]

**Comment by code-mm** [comment]
> Seems like a very useful feature!
> 
> Can you please provide a usage example to the doc-comment, so I can try it out with my code?

There is actually an example in the PR description. Here is an example for your code.

```
pub fn get_book_subject_areas(book_id: i32) -> QueryResult<(Vec<db_models::SubjectArea>, i64)> {
    books_subject_areas::table
        .filter(books_subject_areas::book_id.eq(book_id))
        .inner_join(subject_areas::table)
        .select(subject_areas::all_columns)
        .into_boxed()
        .paginate(1)
        .per_page(10)
        .load_and_count_pages(&get_conn())
}
```

File: helpers/src/db/pagination.rs (hunk 2 above)
Comment (code-mm) [reply helpers/src/db/pagination.rs:59]: Will do.

File: helpers/src/db/pagination.rs (hunk 1 above)
Comment (code-mm) [reply helpers/src/db/pagination.rs:10]: Yes, I think so. The `Paginated` struct contains the query as boxed.

File: helpers/src/db/pagination.rs (hunk 4 above)
Comment (code-mm) [reply helpers/src/db/pagination.rs:36]: No, you don't get all pages here. The `Vec<U>` contains the elements of the current page. The `i64` is the number of pages you have in total.

File: helpers/src/db/pagination.rs (hunk 3 above)
Comment (code-mm) [reply helpers/src/db/pagination.rs:64]: Not really. These methods are implemented in their own query fragments. When you use this trait they are not required anymore. Internally they have the same logic than this trait, just that this one also does the counting and pagination for you.

File: helpers/src/db/pagination.rs (hunk 5 above)
Comment (code-mm) [reply helpers/src/db/pagination.rs:33]: 1. You would also need to make the self parameter mutable. Would be possible, if you find this easier to understand.
2. The struct `Paginated` has no trait `clone`.

File: helpers/src/db/pagination.rs (hunk 4 above)
Comment (Urhengulas) [reply helpers/src/db/pagination.rs:36]: 👍🏾

File: helpers/src/db/pagination.rs (hunk 5 above)
Comment (Urhengulas) [reply helpers/src/db/pagination.rs:33]: 1. true, needs to be mutable
2. clone could be easily derived

In the end it's your decision, but I find the current solution slightly hard to grasp (but maybe I am just not used to that pattern).

File: helpers/src/db/pagination.rs [hunk 6]
```diff
@@ -7,6 +7,19 @@ use diesel::QueryId;
...
+
+```
+<schema-module>::table
+    .into_boxed()
+    .paginate(<page>)
+    .per_page(<records-per-page>)
+    .load_and_count_pages(<database-connection>)
+```
+This returns a tuple `(Vec<T>, i64)` with the records of the requested page as the first and the amount of pages in the second element.
+**/
```
Comment (Urhengulas) [inline helpers/src/db/pagination.rs:21]: I guess this style of comment won't be rendered as doc-comment?

**Review by Urhengulas — COMMENTED** [review]
I am a bit concerned that we make the trait too generic. If we implement it for every Type `T` (what we are currently doing) we could also call `"random string".paginate()`, which doesn't really make sense.

I'd like to limit it to only work for some kind of query (though I am not sure what the type would be for this..)

File: helpers/src/db/pagination.rs (hunk 3 above)
Comment (Urhengulas) [reply helpers/src/db/pagination.rs:81]: But can't we use the methods to simplify the code on our side?
And do sth like (pseudocode)? :
```rust
self.query.limit(&self.per_page);
self.query.offset((self.page - 1) * self.per_page);
```

**Comment by code-mm** [comment]
I added the requirement of `Query` to the trait.

File: helpers/src/db/pagination.rs (hunk 6 above)
Comment (code-mm) [reply helpers/src/db/pagination.rs:21]: When I do `cargo doc` I can see it in the module documentation. Do you mean something else?

File: helpers/src/db/pagination.rs (hunk 3 above)
Comment (code-mm) [reply helpers/src/db/pagination.rs:81]: The `limit` and `offset` are query methods, which only work on queries. The `self.query` contains the query, which we created before the `.paginate`, but we have to execute it as subquery within the query we are building here. The `out` contains the syntax tree of the created query and has the type AstPass, which does not implement them. This part of the QueryFragment is where you have to write SQL. I don't see a way around it.

File: helpers/src/db/pagination.rs (hunk 6 above)
Comment (Urhengulas) [reply helpers/src/db/pagination.rs:21]: Ah okay. Just looked it up.
* no doc-comment
    ```rust
    /* */
    ```
* doc-comment
    ```rust
    /** **/
    ```

**Comment by Urhengulas** [comment]
> I added the requirement of `Query` to the trait.

I don't know what is the right type, but only `Sized` seems to loose...

**Comment by code-mm** [comment]
With the last change the `query` trait enforces to have a diesel query.

**Review by Urhengulas — CHANGES_REQUESTED** [review]
Played around with it. Currently am trying to think about common edge cases we should try to avoid.

**Things I tried**
* set page size to 0
    > Setting page size to 0, results in getting a empty vector (as expected) and a number of pages of 0 (not expected). Should we leave it like that or maybe throw an error?
* call `.limit()` or `.offset()` before `.paginate()`
    > paginate only operates on the restricted set of row (as expected)
* request non-existent page
    > I am having 25 elements, the page size is 10, so I have 3 pages in total. When I am trying to get page 5, I am getting an empty vector (expected) and a number of pages of 0 (not expected). Here I would go with returning either an error, or and empty vector, but with the correct number of pages.

**Comment by Urhengulas** [comment]
I refactored and extended the doc-comment a bit, so it actually runs as a test :)

**Comment by code-mm** [comment]
> Played around with it. Currently am trying to think about common edge cases we should try to avoid.
> 
> **Things I tried**
> 
>     * set page size to 0
>       > Setting page size to 0, results in getting a empty vector (as expected) and a number of pages of 0 (not expected). Should we leave it like that or maybe throw an error?
This behaviour is expected, because when you set the page size to `0`, then you get an empty result, means you miss the count in the database result. Therefore, the method `load_and_count_pages` operates with `0` results.
>     * call `.limit()` or `.offset()` before `.paginate()`
>       > paginate only operates on the restricted set of row (as expected)
This would result in a subquery with limit and offset set. Then the method `load_and_count_pages` gets a count of exactly the maximal amount of results from the subquery.
>     * request non-existent page
>       > I am having 25 elements, the page size is 10, so I have 3 pages in total. When I am trying to get page 5, I am getting an empty vector (expected) and a number of pages of 0 (not expected). Here I would go with returning either an error, or and empty vector, but with the correct number of pages.
This is the same problem as before. Because of the empty result vector you don't have a count value available.

The only way to solve this, without any hackish workarounds, is to have another query to count all the rows and do the calculation based on that. I don't think this is doable in the same call.

**Comment by code-mm** [comment]
I think a solution would be to not allow a `per_page` value lower than `1`. This would result in always requiring at least one row. There would be still the page after last page problem, but this should not really happen... :-D

**Comment by Urhengulas** [comment]
> I think a solution would be to only allow a `per_page` value `>=1`. This would result in always requiring at least one row. Question then is, should we throw an error if the value is too low or just overwrite it with one? Later one, should be easier.
> There would be still the not existing page problem, but this should not really happen... :-D

I'd like to return a 400 http error to the client. Therefore we need a `Error::BadRequest` and return this as a result.

**Review by Urhengulas — APPROVED** [review]
Made some minor adjustions. What do you think about them?

Great work! This will make my life way easier!

File: helpers/src/rpc.rs [hunk 7]
```diff
@@ -16,7 +16,7 @@ impl From<DBError> for Error {
         log::debug!("{}", e);
         match e {
             DBError::NotFound => Error::NotFound,
-            DBError::QueryBuilderError(_) => Error::InvalidData,
+            DBError::QueryBuilderError(_) => Error::InvalidInput,
```
Comment (code-mm) [inline helpers/src/rpc.rs]: I would suggest keeping the other one. I did this because of this: https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.InvalidInput

Comment (Urhengulas) [reply helpers/src/rpc.rs]: Mhh, yeah makes sense. As long as we resolve it in a `400` I am fine with it.
